x/ecocredit/server/orderfill/manager.go

```diff
@@ -0,0 +1,196 @@
+	ecocreditv1beta1 "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1beta1"
+	"github.com/regen-network/regen-ledger/types/math"
+)
+
+// Status represents the order fill status.
+type Status int
+
+const (
+	NotFilled Status = iota
+	BothFilled
+	BuyFilled
+	SellFilled
+)
+
+func (s Status) String() string {
+	switch s {
+	case BothFilled:
+		return "BothFilled"
+	case BuyFilled:
+		return "BuyFilled"
+	case SellFilled:
+		return "SellFilled"
+	default:
+		return "undefined"
+	}
+}
+
+type manager struct {
+	marketplaceStore marketplacev1beta1.StateStore
+	ecocreditStore   ecocreditv1beta1.StateStore
+	bankBalances     map[string]sdk.Int
+	transferManager  TransferManager
+}
+
+// Manager handles order filling for orders already matched by the order book.
+type Manager interface {
+	// Fill fills the provided buy order with the provided sell order and updates
+	// or deletes the orders from the marketplace state. For now, it naively
+	// chooses the bid price as the settlement price.
+	Fill(
+		ctx context.Context,
+		market *marketplacev1beta1.Market,
+		buyOrder *marketplacev1beta1.BuyOrder,
+		sellOrder *marketplacev1beta1.SellOrder,
+	) (Status, error)
+}
+
+// NewManager returns a new Manager.
+func NewManager(db ormdb.ModuleDB, transferManager TransferManager) (Manager, error) {
+	mgr := &manager{transferManager: transferManager}
+
+	var err error
+	mgr.marketplaceStore, err = marketplacev1beta1.NewStateStore(db)
+	if err != nil {
+		return nil, err
+	}
+
+	mgr.ecocreditStore, err = ecocreditv1beta1.NewStateStore(db)
+	if err != nil {
+		return nil, err
+	}
+
+	return mgr, nil
+}
+
+// TransferManager abstracts over transfers of credits and coins. The fill manager
+// is expected to call the transfer manager to handle all transfers rather than
+// attempting to handle them itself.
+type TransferManager interface {
+	// SendCoins sends the specified amount of denom.
+	SendCoins(denom string, amount sdk.Int, from, to sdk.AccAddress) error
+
+	// SendCredits sends the specified amount of credits. If retire is true, the
+	// credits will be retired with the provided retirement location, otherwise
+	// they will be transferred and remain tradable.
+	SendCredits(batchId uint64, amount math.Dec, from, to sdk.AccAddress, retire bool, retirementLocation string) error
+}
+
+func (t manager) Fill(
+	ctx context.Context,
+	market *marketplacev1beta1.Market,
+	buyOrder *marketplacev1beta1.BuyOrder,
+	sellOrder *marketplacev1beta1.SellOrder,
+) (Status, error) {
+	retire := true
+	if buyOrder.DisableAutoRetire {
+		if !sellOrder.DisableAutoRetire {
+			return NotFilled, ecocredit.ErrInvalidBuyOrder.Wrapf("unexpected: auto-retire setting doesn't match, these orders should have never been matched")
+		}
+		retire = false
+	} else {
+		if buyOrder.RetirementLocation == "" {
+			return NotFilled, ecocredit.ErrInvalidBuyOrder.Wrapf("unexpected: retirement location is empty, this order never should have never been created.")
+		}
+	}
```

Comment:
In what situation would these errors be thrown if already checked in the order matching phase?

Reply:
Well... these situations would only arise from errors in matching, so maybe it's not necessary... but it's a good logical check. Maybe we can remove